src/main/java/io/github/opencubicchunks/cubicchunks/core/server/chunkio/RegionCubeStorage.java

```diff
@@ -0,0 +1,248 @@
+        try {
+            //TODO: i'm pretty sure this should be false
+            Optional<ByteBuffer> data = this.save.load(new EntryLocation3D(pos.getX(), pos.getY(), pos.getZ()), true);
+            return data.isPresent()
+                    ? CompressedStreamTools.readCompressed(new ByteArrayInputStream(data.get().array())) //decompress and parse NBT
+                    : null; //cube doesn't exist
+        } finally {
+            this.lock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public void writeColumn(ChunkPos pos, NBTTagCompound nbt) throws IOException {
+        ByteBuf compressedBuf = ByteBufAllocator.DEFAULT.ioBuffer();
+        this.lock.readLock().lock();
+        try {
+            //compress NBT data
+            CompressedStreamTools.writeCompressed(nbt, new ByteBufOutputStream(compressedBuf));
+
+            //write compressed data to disk
+            this.save.save2d(new EntryLocation2D(pos.x, pos.z), compressedBuf.internalNioBuffer(compressedBuf.readerIndex(), compressedBuf.readableBytes()));
+        } finally {
+            this.lock.readLock().unlock();
+            compressedBuf.release();
+        }
+    }
+
+    @Override
+    public void writeCube(CubePos pos, NBTTagCompound nbt) throws IOException {
+        ByteBuf compressedBuf = ByteBufAllocator.DEFAULT.ioBuffer();
+        this.lock.readLock().lock();
+        try {
+            //compress NBT data
+            CompressedStreamTools.writeCompressed(nbt, new ByteBufOutputStream(compressedBuf));
+
+            //write compressed data to disk
+            this.save.save3d(new EntryLocation3D(pos.getX(), pos.getY(), pos.getZ()), compressedBuf.internalNioBuffer(compressedBuf.readerIndex(), compressedBuf.readableBytes()));
+        } finally {
+            this.lock.readLock().unlock();
+            compressedBuf.release();
+        }
+    }
+
+    @Override
+    public void forEachColumn(Consumer<ChunkPos> callback) throws IOException {
+        this.lock.readLock().lock();
+        try {
+            this.save.getSaveSection2D().forAllKeys(pos -> callback.accept(new ChunkPos(pos.getEntryX(), pos.getEntryZ())));
+        } finally {
+            this.lock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public void forEachCube(Consumer<CubePos> callback) throws IOException {
+        this.lock.readLock().lock();
+        try {
+            this.save.getSaveSection3D().forAllKeys(pos -> callback.accept(new CubePos(pos.getEntryX(), pos.getEntryY(), pos.getEntryZ())));
+        } finally {
+            this.lock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public void flush() throws IOException {
+        this.lock.writeLock().lock();
+        try {
+            //close and re-open save
+            //this will also close all the region files, which should ensure that their contents are written to disk
+            this.save.close();
+            this.save = saveForPath(this.path);
```

Comment:
Now that flush is merged, can you update to use it?

Reply:
yup